src/rust/engine/src/externs.rs

```diff
@@ -489,14 +480,24 @@ pub enum GeneratorResponse {
 }
 
 ///
-/// The result of an `identify` call, including the __hash__ of a Handle and its TypeId.
+/// We need to make sure each type object is registered in `ExternContext#_types` in native.py by
+/// running it through an extern before wrapping it in a `TypeId`, so we codify that by wrapping it
+/// in this struct after going through a `product_type` call.
 ///
 #[repr(C)]
-pub struct Ident {
+pub struct ProductTypeId {
   pub hash: i64,
```

Comment:
It probably isn't strictly necessary to have a hash on `ProductTypeId`... it should be ok to just use the derived hashcode (based on hashing the TypeId, which is distinct).

And given that, I'm not sure that `ProductTypeId` needs to exist... could just be `TypeId`?

Reply:
Was able to convert this to just `TypeId` by hashing it manually -- let me know if the `PRODUCT_TYPE_ID_HASH_BUILDER` lazy static in `interns.rs` looks right: https://github.com/pantsbuild/pants/pull/7114/files#diff-8ab5c4db2ca5c0325da0cb5aecefaff7R43

Reply:
I meant just `#[derive(Hash)]`, I think. But I might have been missing something. Regardless, not worth another CI burn.

Reply:
Oh...well `TypeId` already derives `Hash`? This was just my best attempt at turning the `TypeId` into an `i64` for the `InternKey`, and since we were already using the `FNV` hash it seemed reasonable to use again. I'm not sure if the `as i64` affects any of the hashing properties, or if there's an easier way to get an `i64` from an instance of a hashable type.